File: LEDAppVBProject/LEDServerVBApp/FormMainServer.vb
Imports System.Net.Sockets
Imports System.Net
Imports System.Threading

Public Class FormMainServer
    Shared clients As New List(Of TcpClient)
    Shared server As TcpListener
    Shared ReadOnly serverLock As Object = New Object()
    Shared serverAvailable As Boolean = False


    Public Sub New()

        ' This call is required by the designer.
        InitializeComponent()

        ' Add any initialization after the InitializeComponent() call.
        setInitLEDColors()

        'Gets the IP Address
        Dim hostName As String = Dns.GetHostName() ' Retrive the Name of HOST            
        Dim hostnameIP As IPHostEntry = Dns.GetHostEntry(hostName)
        Dim ip() As IPAddress = hostnameIP.AddressList
        For Each item As IPAddress In ip

            If (item.AddressFamily = AddressFamily.InterNetwork) Then
                textHostIPAddress.Text = item.ToString()
            End If
        Next
    End Sub

    Private Sub setInitLEDColors()
        leftLED0.BackColor = Color.Red
        leftLED1.BackColor = Color.Blue
        leftLED2.BackColor = Color.Green
        leftLED3.BackColor = Color.Orange
        rightLED0.BackColor = Color.Black
        rightLED1.BackColor = Color.Black
        rightLED2.BackColor = Color.Black
        rightLED3.BackColor = Color.Black
    End Sub


    Private Sub buttonConnect_Click(sender As Object, e As EventArgs) Handles buttonConnect.Click
        Try

            Dim hostIP As String = textHostIPAddress.Text.Trim()
            Dim portNo As Integer = Convert.ToInt32(textPortNo.Text.Trim())
            server = New TcpListener(IPAddress.Parse(hostIP), portNo)
            server.Start()

            serverAvailable = True
            'Starting the Client listening to any incoming Data from Server.
            Dim t As Thread = New Thread(AddressOf Me.addIncomingClients)
            t.IsBackground = True
            t.Start()

            buttonConnect.Text = "Running"
        Catch ex As Exception
            Console.WriteLine("FormMainServer.buttonConnect+Click() " + ex.Message)
            serverAvailable = False
        End Try
    End Sub

    Private Sub addIncomingClients()

        While (serverAvailable)
            Try

                Dim client As TcpClient = server.AcceptTcpClient()
                'Let the client object listen for any incoming              
                Dim t As Thread = New Thread(AddressOf Me.listenToClientForData)
                t.IsBackground = True
                t.Start(client)

                SyncLock (serverLock)
                    clients.Add(client)
                End SyncLock
            Catch e As Exception
                Console.WriteLine("FormMainServer.addIncomingClients()" + e.Message)
            End Try
        End While
    End Sub

    Private Sub listenToClientForData(pclient As Object)

        Dim client As TcpClient
        Dim stream As NetworkStream

        Try

            SyncLock (serverLock)

                client = CType(pclient, TcpClient)
                stream = client.GetStream()
            End SyncLock

            While (True)


                SyncLock (serverLock)

                    If (stream.DataAvailable) Then

                        Dim mData(3) As Byte
                        stream.Read(mData, 0, mData.Length)                      
                        Select Case (mData(0))

                            Case 1 'Set LED
                                setLEDs(mData)
                                'Updates all Clients
                                sendDataToAllClients()
                                Exit Select
                            Case 2 'Get LED
                                sendDataToClient(client)
                                Exit Select
                        End Select
                    End If

                End SyncLock
            End While

        Catch e As Exception
            Console.WriteLine("FormMainServer.listenToClientForData()" + e.Message)
        End Try
    End Sub

    Private Sub setLEDs(mData As Byte())

        'Setting Left LEDs
        setValue((mData(1) And &H1) > 0, 0)
        setValue((mData(1) And &H2) > 0, 1)
        setValue((mData(1) And &H4) > 0, 2)
        setValue((mData(1) And &H8) > 0, 3)

        'Setting Right LEDs
        setValue((mData(2) And &H1) > 0, 4)
        setValue((mData(2) And &H2) > 0, 5)
        setValue((mData(2) And &H4) > 0, 6)
        setValue((mData(2) And &H8) > 0, 7)
    End Sub

    Delegate Sub SetValueCallback(b As Boolean, control As Integer)

    Private Sub setValue(b As Boolean, control As Integer)
        ' InvokeRequired required compares the thread ID of the
        ' calling thread to the thread ID of the creating thread.
        ' If these threads are different, it returns true.
        Select Case (control)

            Case 0
                If (leftLED0.InvokeRequired) Then

                    Dim d As SetValueCallback = New SetValueCallback(AddressOf setValue)
                    Me.Invoke(d, New Object() {b, control})

                Else

                    leftLED0.BackColor = IIf(b, Color.Red, Color.Black)
                End If
                Exit Select
            Case 1
                If (leftLED1.InvokeRequired) Then

                    Dim d As SetValueCallback = New SetValueCallback(AddressOf setValue)
                    Me.Invoke(d, New Object() {b, control})

                Else

                    leftLED1.BackColor = IIf(b, Color.Blue, Color.Black)
                End If
                Exit Select
            Case 2
                If (leftLED2.InvokeRequired) Then

                    Dim d As SetValueCallback = New SetValueCallback(AddressOf setValue)
                    Me.Invoke(d, New Object() {b, control})

                Else

                    leftLED2.BackColor = IIf(b, Color.Green, Color.Black)
                End If
                Exit Select
            Case 3
                If (leftLED3.InvokeRequired) Then

                    Dim d As SetValueCallback = New SetValueCallback(AddressOf setValue)
                    Me.Invoke(d, New Object() {b, control})

                Else

                    leftLED3.BackColor = IIf(b, Color.Orange, Color.Black)
                End If
                Exit Select
            Case 4
                If (rightLED0.InvokeRequired) Then

                    Dim d As SetValueCallback = New SetValueCallback(AddressOf setValue)
                    Me.Invoke(d, New Object() {b, control})

                Else

                    rightLED0.BackColor = IIf(b, Color.Red, Color.Black)
                End If
                Exit Select
            Case 5
                If (rightLED1.InvokeRequired) Then

                    Dim d As SetValueCallback = New SetValueCallback(AddressOf setValue)
                    Me.Invoke(d, New Object() {b, control})

                Else

                    rightLED1.BackColor = IIf(b, Color.Blue, Color.Black)
                End If
                Exit Select
            Case 6
                If (rightLED2.InvokeRequired) Then

                    Dim d As SetValueCallback = New SetValueCallback(AddressOf setValue)
                    Me.Invoke(d, New Object() {b, control})

                Else

                    rightLED2.BackColor = IIf(b, Color.Green, Color.Black)
                End If
                Exit Select
            Case 7
                If (rightLED3.InvokeRequired) Then

                    Dim d As SetValueCallback = New SetValueCallback(AddressOf setValue)
                    Me.Invoke(d, New Object() {b, control})

                Else
                    rightLED3.BackColor = IIf(b, Color.Yellow, Color.Black)
                End If
                Exit Select
        End Select

    End Sub

    Private Sub sendDataToClient(ByVal pclient As Object)

        Dim client As TcpClient
        Dim stream As NetworkStream

        Try

            SyncLock (serverLock)

                client = CType(pclient, TcpClient)
                stream = client.GetStream()
                'Sending Back the current LED Status
                Dim data() As Byte = getByteForGETLEDForClients()
                stream.Write(data, 0, data.Length)
            End SyncLock

        Catch e As Exception

            Console.WriteLine("FormMainServer.sendDataToClient()" + e.Message)
        End Try

    End Sub

    Private Sub sendDataToAllClients()

        SyncLock (serverLock)

            For i As Integer = 0 To clients.Count - 1

                Try
                    Dim client As TcpClient = clients.ElementAt(i)
                    Dim stream As NetworkStream = client.GetStream()
                    'Sending Back the current LED Status
                    Dim mData() As Byte = getByteForGETLEDForClients()
                    stream.Write(mData, 0, mData.Length)                    
                Catch e As Exception
                    Console.WriteLine("FormMainServer.sendDataToAllClients()" + e.Message)
                End Try

            Next
        End SyncLock
    End Sub

    Private Function getByteForGETLEDForClients() As Byte()

        Dim mData(3) As Byte
        mData(0) = 3 'Command
        mData(1) = &H0
        mData(2) = &H0

        'Setting Left LEDS
        Dim val As Byte = IIf(leftLED0.BackColor = Color.Red, &H1, &H0)
        mData(1) = (mData(1) Or val)
        val = IIf(leftLED1.BackColor = Color.Blue, &H2, &H0)
        mData(1) = (mData(1) Or val)
        val = IIf(leftLED2.BackColor = Color.Green, &H4, &H0)
        mData(1) = (mData(1) Or val)
        val = IIf(leftLED3.BackColor = Color.Orange, &H8, &H0)
        mData(1) = (mData(1) Or val)
        'Setting Right LEDS
        val = IIf(rightLED0.BackColor = Color.Red, &H1, &H0)
        mData(2) = (mData(2) Or val)
        val = IIf(rightLED1.BackColor = Color.Blue, &H2, &H0)
        mData(2) = (mData(2) Or val)
        val = IIf(rightLED2.BackColor = Color.Green, &H4, &H0)
        mData(2) = (mData(2) Or val)
        val = IIf(rightLED3.BackColor = Color.Yellow, &H8, &H0)
        mData(2) = (mData(2) Or val)

        'Returning the value        
        getByteForGETLEDForClients = mData
    End Function

    Private Sub leftLED0_Click(sender As Object, e As EventArgs) Handles leftLED0.Click
        'Toggle Back Color of the button
        leftLED0.BackColor = IIf(leftLED0.BackColor = Color.Black, Color.Red, Color.Black)
    End Sub

    Private Sub leftLED1_Click(sender As Object, e As EventArgs) Handles leftLED1.Click
        'Toggle Back Color of the button
        leftLED1.BackColor = IIf(leftLED1.BackColor = Color.Black, Color.Blue, Color.Black)
    End Sub


    Private Sub leftLED2_Click(sender As Object, e As EventArgs) Handles leftLED2.Click
        'Toggle Back Color of the button
        leftLED2.BackColor = IIf(leftLED2.BackColor = Color.Black, Color.Green, Color.Black)
    End Sub

    Private Sub leftLED3_Click(sender As Object, e As EventArgs) Handles leftLED3.Click
        'Toggle Back Color of the button
        leftLED3.BackColor = IIf(leftLED3.BackColor = Color.Black, Color.Orange, Color.Black)
    End Sub


    Private Sub rightLED0_Click(sender As Object, e As EventArgs) Handles rightLED0.Click
        'Toggle Back Color of the button
        rightLED0.BackColor = IIf(rightLED0.BackColor = Color.Black, Color.Red, Color.Black)
    End Sub

    Private Sub rightLED1_Click(sender As Object, e As EventArgs) Handles rightLED1.Click
        'Toggle Back Color of the button
        rightLED1.BackColor = IIf(rightLED1.BackColor = Color.Black, Color.Blue, Color.Black)
    End Sub

    Private Sub rightLED2_Click(sender As Object, e As EventArgs) Handles rightLED2.Click
        'Toggle Back Color of the button
        rightLED2.BackColor = IIf(rightLED2.BackColor = Color.Black, Color.Green, Color.Black)
    End Sub

    Private Sub rightLED3_Click(sender As Object, e As EventArgs) Handles rightLED3.Click
        'Toggle Back Color of the button
        rightLED3.BackColor = IIf(rightLED3.BackColor = Color.Black, Color.Yellow, Color.Black)
    End Sub
End Class

File: LEDAppC#Project/LEDServerApp/FormMainServer.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace LEDServerApp
{
    public partial class FormMainServer : Form
    {
        static List<TcpClient> clients = new List<TcpClient>(); 
        static TcpListener server;        
        static readonly Object serverLock = new Object();
        static bool serverAvailable = false;
        public FormMainServer()
        {
            InitializeComponent();
            setInitLEDColors();

            //Gets the IP Address
            string hostName = Dns.GetHostName(); // Retrive the Name of HOST            
            IPHostEntry hostname = Dns.GetHostEntry(hostName);
            IPAddress[] ip = hostname.AddressList;
            foreach (IPAddress item in ip)
            {
                if(item.AddressFamily == AddressFamily.InterNetwork){
                    textHostIPAddress.Text = item.ToString();
                }                                
            }
            
        }

        private void setInitLEDColors()
        {
            leftLED0.BackColor = Color.Red;
            leftLED1.BackColor = Color.Blue;
            leftLED2.BackColor = Color.Green;
            leftLED3.BackColor = Color.Orange;
            rightLED0.BackColor = Color.Black;
            rightLED1.BackColor = Color.Black;
            rightLED2.BackColor = Color.Black;
            rightLED3.BackColor = Color.Black;
        }

        private void buttonConnect_Click(object sender, EventArgs e)
        {
            try
            {
                String hostIP = textHostIPAddress.Text.Trim();                 
                int portNo = Convert.ToInt32(textPortNo.Text.Trim());
                server = new TcpListener(IPAddress.Parse(hostIP),portNo);
                server.Start();

                serverAvailable = true;
                //Starting the Client listening to any incoming Data from Server.
                Thread t=new Thread(addIncomingClients);
                t.IsBackground = true;
                t.Start();

                buttonConnect.Text = "Running";
            }catch(Exception ex){
                Console.WriteLine("FormMainServer.buttonConnect+Click() "+ex.Message);
                serverAvailable = false;
            }
            
        }

        private void addIncomingClients()
        {
            while(serverAvailable){
                try
                {
                    TcpClient client = server.AcceptTcpClient();
                    //Let the client object listen for any incoming              
                    Thread t = new Thread(listenToClientForData);
                    t.IsBackground = true;
                    t.Start(client);

                    lock (serverLock)
                    {
                        clients.Add(client);
                    }                
                }
                catch (Exception e)
                {
                    Console.WriteLine("FormMainServer.addIncomingClients()" + e.Message);
                }
                
            }
        }

        private void listenToClientForData(Object pclient)
        {
            TcpClient client;
            NetworkStream stream;

            try
            {
                lock (serverLock)
                {
                    client = (TcpClient)pclient;
                    stream = client.GetStream();
                }

                while (true)
                {

                    lock (serverLock)
                    {
                        if (stream.DataAvailable)
                        {
                            byte[] data = new byte[3];
                            stream.Read(data, 0, data.Length);
                            switch (data[0])
                            {
                                case 1: //Set LED
                                    setLEDs(data);
                                    //Updates all Clients
                                    sendDataToAllClients();
                                    break;
                                case 2: //Get LED
                                    sendDataToClient(client);
                                    break;
                                default:
                                    break;
                            }
                        }

                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("FormMainServer.listenToClientForData()" + e.Message);
            }
            
        }

        private void setLEDs(byte[] data)
        {
            //Setting Left LEDs
            setValue((data[1] & 0X01) > 0,0);
            setValue((data[1] & 0X02) > 0,1);
            setValue((data[1] & 0X04) > 0,2);
            setValue((data[1] & 0X08) > 0,3);

            //Setting Right LEDs
            setValue((data[2] & 0X01) > 0,4);
            setValue((data[2] & 0X02) > 0,5);
            setValue((data[2] & 0X04) > 0,6);
            setValue((data[2] & 0X08) > 0,7);
        }

        delegate void SetValueCallback(bool b,int control);

        private void setValue(bool b,int control)
        {
            // InvokeRequired required compares the thread ID of the
            // calling thread to the thread ID of the creating thread.
            // If these threads are different, it returns true.
            switch (control)
            {
                case 0:
                    if (leftLED0.InvokeRequired)
                    {
                        SetValueCallback d = new SetValueCallback(setValue);
                        this.Invoke(d, new object[] { b, control });
                    }
                    else
                    {
                        leftLED0.BackColor = b ? Color.Red : Color.Black;
                    }
                    break;
                case 1:
                    if (leftLED1.InvokeRequired)
                    {
                        SetValueCallback d = new SetValueCallback(setValue);
                        this.Invoke(d, new object[] { b, control });
                    }
                    else
                    {
                        leftLED1.BackColor = b ? Color.Blue : Color.Black;
                    }
                    break;
                case 2:
                    if (leftLED2.InvokeRequired)
                    {
                        SetValueCallback d = new SetValueCallback(setValue);
                        this.Invoke(d, new object[] { b, control });
                    }
                    else
                    {
                        leftLED2.BackColor = b ? Color.Green : Color.Black;
                    }
                    break;
                case 3:
                    if (leftLED3.InvokeRequired)
                    {
                        SetValueCallback d = new SetValueCallback(setValue);
                        this.Invoke(d, new object[] { b, control });
                    }
                    else
                    {
                        leftLED3.BackColor = b ? Color.Orange : Color.Black;
                    }
                    break;
                case 4:
                    if (rightLED0.InvokeRequired)
                    {
                        SetValueCallback d = new SetValueCallback(setValue);
                        this.Invoke(d, new object[] { b, control });
                    }
                    else
                    {
                        rightLED0.BackColor = b ? Color.Red : Color.Black;
                    }
                    break;
                case 5:
                    if (rightLED1.InvokeRequired)
                    {
                        SetValueCallback d = new SetValueCallback(setValue);
                        this.Invoke(d, new object[] { b, control });
                    }
                    else
                    {
                        rightLED1.BackColor = b ? Color.Blue : Color.Black;
                    }
                    break;
                case 6:
                    if (rightLED2.InvokeRequired)
                    {
                        SetValueCallback d = new SetValueCallback(setValue);
                        this.Invoke(d, new object[] { b, control });
                    }
                    else
                    {
                        rightLED2.BackColor = b ? Color.Green : Color.Black;
                    }
                    break;
                case 7:
                    if (rightLED3.InvokeRequired)
                    {
                        SetValueCallback d = new SetValueCallback(setValue);
                        this.Invoke(d, new object[] { b, control });
                    }
                    else
                    {
                        rightLED3.BackColor = b ? Color.Yellow : Color.Black;
                    }
                    break;
            }
            
        }

        private void sendDataToClient(Object pclient)
        {
            TcpClient client = (TcpClient)pclient;
            NetworkStream stream;

            try
            {
                lock (serverLock)
                {
                    client = (TcpClient)pclient;
                    stream = client.GetStream();
                    //Sending Back the current LED Status
                    byte[] data = getByteForGETLEDForClients();
                    stream.Write(data, 0, data.Length);
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("FormMainServer.sendDataToClient()" + e.Message);
            }
            
        }

        private void sendDataToAllClients()
        {            
            lock (serverLock)
            {
                for (int i = 0; i < clients.Count; i++)
                {
                    try
                    {
                        TcpClient client = clients.ElementAt(i);
                        NetworkStream stream = client.GetStream();
                        //Sending Back the current LED Status
                        byte[] data = getByteForGETLEDForClients();
                        stream.Write(data, 0, data.Length);
                    }
                    catch (Exception e)
                    {
                        Console.WriteLine("FormMainServer.sendDataToAllClients()" + e.Message);
                    }
                    
                }                
            }
        }

        private byte[] getByteForGETLEDForClients()
        {
            byte[] data = new byte[3];
            data[0] = 3;//Command
            data[1] = 0X00;
            data[2] = 0X00;
            //Setting Left LEDS
            byte val = (byte)(leftLED0.BackColor == Color.Red ? 0X01 : 0X00);
            data[1] = (byte)(data[1] | val);
            val = (byte)(leftLED1.BackColor == Color.Blue ? 0X02 : 0X00);
            data[1] = (byte)(data[1] | val);
            val = (byte)(leftLED2.BackColor == Color.Green ? 0X04 : 0X00);
            data[1] = (byte)(data[1] | val);
            val = (byte)(leftLED3.BackColor == Color.Orange ? 0X08 : 0X00);
            data[1] = (byte)(data[1] | val);
            //Setting Right LEDS
            val = (byte)(rightLED0.BackColor == Color.Red ? 0X01 : 0X00);
            data[2] = (byte)(data[2] | val);
            val = (byte)(rightLED1.BackColor == Color.Blue ? 0X02 : 0X00);
            data[2] = (byte)(data[2] | val);
            val = (byte)(rightLED2.BackColor == Color.Green ? 0X04 : 0X00);
            data[2] = (byte)(data[2] | val);
            val = (byte)(rightLED3.BackColor == Color.Yellow  ? 0X08 : 0X00);
            data[2] = (byte)(data[2] | val);
            return data;
        }

        private void leftLED0_Click(object sender, EventArgs e)
        {
            //Toggle Back Color of the button
            leftLED0.BackColor = leftLED0.BackColor == Color.Black ? Color.Red : Color.Black;
        }

        private void leftLED1_Click(object sender, EventArgs e)
        {
            //Toggle Back Color of the button
            leftLED1.BackColor = leftLED1.BackColor == Color.Black ? Color.Blue : Color.Black;
        }

        private void leftLED2_Click(object sender, EventArgs e)
        {
            //Toggle Back Color of the button
            leftLED2.BackColor = leftLED2.BackColor == Color.Black ? Color.Green : Color.Black;
        }

        private void leftLED3_Click(object sender, EventArgs e)
        {
            //Toggle Back Color of the button
            leftLED3.BackColor = leftLED3.BackColor == Color.Black ? Color.Orange : Color.Black;
        }

        private void rightLED0_Click(object sender, EventArgs e)
        {
            //Toggle Back Color of the button
            rightLED0.BackColor = rightLED0.BackColor == Color.Black ? Color.Red : Color.Black;
        }

        private void rightLED1_Click(object sender, EventArgs e)
        {
            //Toggle Back Color of the button
            rightLED1.BackColor = rightLED1.BackColor == Color.Black ? Color.Blue : Color.Black;
        }

        private void rightLED2_Click(object sender, EventArgs e)
        {
            //Toggle Back Color of the button
            rightLED2.BackColor = rightLED2.BackColor == Color.Black ? Color.Green : Color.Black;
        }

        private void rightLED3_Click(object sender, EventArgs e)
        {
            //Toggle Back Color of the button
            rightLED3.BackColor = rightLED3.BackColor == Color.Black ? Color.Yellow : Color.Black;
        }
        
    }
}
